rustls/examples/internal/bogo_shim.rs

```diff
@@ -198,8 +198,8 @@ impl server::ClientCertVerifier for DummyClientAuth {
         self.mandatory
     }
 
-    fn client_auth_root_subjects(&self) -> DistinguishedNames {
-        DistinguishedNames::new()
+    fn client_auth_root_subjects(&self) -> &[Vec<u8>] {
```

Comment:
I don't think this is a good idea.

First, my experience from benchmarking #1179 showed that allocations don't matter all that much in the performance of our TLS stack, and this is just a single allocation. To me, it seems perfectly valid for implementers of this trait to want to yield a list of subjects that is not tied to the lifetime of the verifier itself.

Second, I think going from the semantically useful `DistinguishedNames` return type to the completely generic `&[Vec<u8>]` seems to be a step back.

Reply:
Agreed on the semantically useful name. And I like #1245 a lot. I see that it makes `DistinguishedName` a struct with private fields, so avoids exposing the `PayloadU8` type _and_ the `VecU16OfPayloadU8` type. 👍🏻

I agree in general with not prematurely optimizing on allocations, but I think it's worth avoiding public interfaces that _prevent_ us from optimizing on allocations if that later becomes something we want.

> To me, it seems perfectly valid for implementers of this trait to want to yield a list of subjects that is not tied to the lifetime of the verifier itself.

In general, yes. For the specific case of returning a list of root subjects for a client cert verifier, I think it's extremely unlikely. Roots (whether for client certificates or for server certificates) change rather infrequently. Also, with the change that remove the `sni` parameter from this call, any moderately complex behavior from this trait involves creating (or selecting) a different ClientCertVerifier with each handshake. So it's not burdensome for the lifetime of a `&[DistinguishedName]` to be tied to a ClientCertVerifier.

Reply:
> I agree in general with not prematurely optimizing on allocations, but I think it's worth avoiding public interfaces that _prevent_ us from optimizing on allocations if that later becomes something we want.

While I think we're in agreement, I just want to highlight that this is not just "not prematurely optimizing": in the PR linked above I avoided a zillion alllocations on the handshake path and performance was not noticeably faster (and seemed to be worse at times). So it looks to me like, in the context of the TLS handshake, most (small) allocations largely do not matter, and simplifying along other axes is more important.

> Also, with the change that remove the `sni` parameter from this call, any moderately complex behavior from this trait involves creating (or selecting) a different ClientCertVerifier with each handshake. So it's not burdensome for the lifetime of a `&[DistinguishedName]` to be tied to a ClientCertVerifier.

This is fair. However, me might want to reconsider how we document this, since we still say "Making one of these can be expensive, and should be once per process rather than once per connection." for `ServerConfig` and `ClientConfig`, and we might want to nuance that going forward to suggest that, especially on the server side, doing so is probably not overly expensive (I think the main expensive part these days is populating the client's `RootCertStore`?). This comes up somewhat regularly in issues...